api/org/commcare/api/json/JsonActionUtils.java

```diff
@@ -0,0 +1,262 @@
+            }
+            return new SelectMultiData(ret);
+        } else if (formEntryPrompt.getDataType() == Constants.DATATYPE_GEOPOINT){
+            return AnswerDataFactory.template(formEntryPrompt.getControlType(), formEntryPrompt.getDataType()).cast(
+                    new UncastData(data.replace(",", " ").replace("[","").replace("]", "")));
+        }
+
+        return data.equals("") ? null : AnswerDataFactory.template(formEntryPrompt.getControlType(), formEntryPrompt.getDataType()).cast(new UncastData(data));
+    }
+
+    /**
+     * OK, this function is kind of a monster. Given a FormDef and a String representation of the form index,
+     * return a full fledged FormIndex object.
+     */
+    private static FormIndex indexFromString(String stringIndex, FormDef form){
+        if(stringIndex == null || stringIndex.equals("None")){
+            return null;
+        } else if(stringIndex.equals("<")){
+            return FormIndex.createBeginningOfFormIndex();
+        } else if(stringIndex.equals(">")){
+            return FormIndex.createEndOfFormIndex();
+        }
+
+        List<Pair<Integer, Integer>> list = stepToList(stringIndex);
+
+        FormIndex ret = reduceFormIndex(list, null);
+        ret.assignRefs(form);
+        return ret;
+    }
+
+    /**
+     * Given a String represnetation of a form index, decompose it into a list of <index, multiplicity> pairs
+     * @param index the comma separated String representation of the form index
+     * @return @index represented as a list of index,multiplicity integer pairs
+     */
+    private static List<Pair<Integer, Integer>> stepToList(String index){
+        ArrayList<Pair<Integer, Integer>> ret = new ArrayList<>();
+        String[] split = index.split(",");
+        List<String> list = Arrays.asList(split);
+        Collections.reverse(list);
+        for(String step: list){
+            if(!step.trim().equals("")) {
+                Pair<Integer, Integer> pair = stepFromString(step);
+                ret.add(pair);
+            }
+        }
+        return ret;
+    }
+
+    /**
+     * Given the string representation of one "Step" in a form index, return an Integer pair of <index, multiplicity>
+     */
+    private static Pair<Integer, Integer> stepFromString(String step){
+        // honestly not sure. thanks obama/drew
+        if(step.endsWith("J")){
+            return new Pair<>(Integer.parseInt("" + step.substring(0, step.length()-1)), -10);
+        }
+        // we want to deal with both '.' and '_' as separators for the time being for TF legacy reasons
+        String[] split = step.split("[._:]");
+        // the form index is the first part, the multiplicity is the second
+        int i = Integer.parseInt(split[0].trim());
+        int mult = -1;
+        try{
+            mult = Integer.parseInt(split[1].trim());
+        } catch(IndexOutOfBoundsException | NullPointerException e){
+            // do nothing, leave mult as -1
+        }
```

Comment:
I prefer an `if (split.length > 1 && split[1] != null) {...` instead of a `try` / `catch`